Option to include '@' symbol as part of a link

rylnd opened the pull request:
Currently, `auto_link_usernames_or_lists` places the `@` symbol outside of `<a>` tags.

To more closely match twitter's auto-linking behavior for usernames and lists, this patch adds the option `:include_symbol` that allows `@` to be included as part of the link, rather than before it.

This patch also includes some fixes to `spec_helper.rb` to prevent errors on failing specs; if needed that could be submitted as a separate patch.

Commit:
Upgraded to the latest conformance suite

Commit:
Fix errors in RSpec failure messages

TypeErrors and NoMethodErrors were being thrown in the case of a failing
spec. This fixes that, along with making the failure messages a bit more
uniform, informative, and readable.

Commit:
Add option to include '@' within a link

Passing :include_symbol => true to auto_link will place the '@' symbol
within the <a> tag instead of outside of it.  This more closely mirrors
the expected/Twitter's behavior for linking usernames and lists.

keitaf commented on lib/autolink.rb:
```diff
@@ -47,6 +47,7 @@ def html_escape(text)
     # <tt>:username_url_base</tt>::      the value for <tt>href</tt> attribute on username links. The <tt>@username</tt> (minus the <tt>@</tt>) will be appended at the end of this.
     # <tt>:list_url_base</tt>::      the value for <tt>href</tt> attribute on list links. The <tt>@username/list</tt> (minus the <tt>@</tt>) will be appended at the end of this.
     # <tt>:hashtag_url_base</tt>::      the value for <tt>href</tt> attribute on hashtag links. The <tt>#hashtag</tt> (minus the <tt>#</tt>) will be appended at the end of this.
+    # <tt>:include_symbol</tt>::    place the <tt>@</tt> symbol within the link.
```
Can we change its name to "username_include_symbol" or something to clarify that this option is only for username/@mention?

rylnd replied:
Sure thing.  I saw you're using both username_ and list_ as prefixes, and this applies to both. 

It could be something more generic like `include_at_symbol` or more specific like `username_and_list_include_symbol`.

Since hashtags already exhibit this behavior, using this option would essentially include _all_ symbols, so I'm leaning toward a more general name.

Thoughts?

rylnd commented:
I'm thinking that `username_include_symbol` would probably be the best of both worlds, unless people want different behavior for usernames and lists.

Commit:
Change option name to 'username_include_symbol'

keitaf commented:
Yes I think 'username_include_symbol' is a good one.

I ported your change to JS and Java. Can you please check them?
https://github.com/twitter/twitter-text-java/pull/16
https://github.com/twitter/twitter-text-js/pull/30

rylnd commented:
@keitaf Those look great.  Thank you for porting those.

Your test suites give me something to aspire to.

Commit:
Upgraded to the latest conformance suite

Commit:
Allow '@' symbol in URL for :username_include_symbol

Refactor since 'twitter.com/@username' is a perfectly valid URL.

Commit:
Don't include '@' sign in username href

When using :username_include_symbol, a link's href does not include the '@' symbol, while its innerText does.

Partially reverts the previous commit.

rylnd commented:
Is this in an acceptable state for pulling?

keitaf commented:
This looks good to me. Let me merge into master.
